disass.ml

```diff
@@ -0,0 +1,45 @@
+open Instr
+
+let disassemble (prog : (Scope.scope_annotation option * Instr.instruction) array) =
+  let lit_to_str lit =
+    match lit with
+    | Int i  -> string_of_int i
+    | Bool b -> string_of_bool b
+    | Nil    -> "nil"
+  in
+  let dump_instr buf (instr_annot : (Scope.scope_annotation option * Instr.instruction)) =
+    let pr = Printf.bprintf in
+    let dump_expr exp =
+      match exp with
+      | Var v             -> pr buf "%s" v
+      | Lit lit           -> pr buf "%s" (lit_to_str lit)
+      | Op (Plus, [a; b]) -> pr buf "(%s + %s)" a b
+      | Op (Eq,   [a; b]) -> pr buf "(%s == %s)" a b
+      | Op (_, _)         -> assert(false)
+    in
+    match instr_annot with
+    | (annot, instr) ->
+        let str_from_vars vars = String.concat ", " (Scope.VarSet.elements vars) in
+        begin match annot with
+        | Some (Scope.Exact e)    -> pr buf "{%s} "      (str_from_vars e)
+        | Some (Scope.At_least e) -> pr buf "{%s, ...} " (str_from_vars e)
+        | None -> ()
+        end;
+        begin match instr with
+        | Decl_const (var, exp)           -> pr buf " const %s = " var; dump_expr exp
+        | Decl_mut (var, exp)             -> pr buf " mut %s = " var; dump_expr exp
+        | Assign (var, exp)               -> pr buf " %s <- " var; dump_expr exp
+        | Branch (exp, l1, l2)            -> pr buf " branch "; dump_expr exp; pr buf " %s %s" l1 l2
+        | Label label                     -> pr buf "%s:" label
+        | Goto label                      -> pr buf " goto %s" label
+        | Print exp                       -> pr buf " print "; dump_expr exp
+        | Invalidate (exp, l, vars)       -> pr buf " invalidate "; dump_expr exp;
+                                             pr buf " %s (%s)" l (String.concat ", " vars)
+        | Stop                            -> pr buf " stop"
+        | Comment str                     -> pr buf " %s" str
```

Comment:
Do we keep the `#` in the comment string? We probably should not. Again that sounds like a parser change, but it would also require a change here.